mteb/evaluation/evaluators/model_encode.py

```diff
@@ -6,35 +6,56 @@
 import numpy as np
 import torch
 
-from mteb.encoder_interface import Encoder
+from mteb.encoder_interface import Encoder, PromptType
 
 logger = logging.getLogger(__name__)
 
 
 def model_encode(
-    sentences: Sequence[str], *, model: Encoder, prompt_name: str | None, **kwargs
+    sentences: Sequence[str],
+    *,
+    model: Encoder,
+    task_name: str | None,
+    prompt_type: PromptType | None = None,
+    **kwargs,
 ) -> np.ndarray:
     """A wrapper function around the model.encode method that handles the prompt_name argument and standardizes the output to a numpy array.
 
     Args:
         sentences: The sentences to encode
         model: The model to use for encoding
-        prompt_name: The prompt name to use for encoding
+        task_name: The task name to use for building the encoding prompt
+        prompt_type: The prompt type (e.g. "query" | "passage") to use for building the encoding prompt
         **kwargs: Additional arguments to pass to the model.encode method
     """
-    kwargs["prompt_name"] = prompt_name
+    ## The order of priorities are:
+    # 1. Composed prompt of task name + prompt type
+    # 2. Specific task prompt
+    # 3. Specific prompt type
+
     if hasattr(model, "prompts"):
         # check if prompts is an empty dict
         if not model.prompts:  # type: ignore
             logger.info(
                 "Model does not support prompts. Removing prompt_name argument."
             )
-            kwargs.pop("prompt_name")
-        elif prompt_name not in model.prompts:  # type: ignore
+            kwargs.pop("prompt_name", None)
+
+        if task_name and prompt_type and f"{task_name}-{ prompt_type}" in model.prompts:
+            kwargs["prompt_name"] = f"{task_name}-{ prompt_type}"
+        elif task_name and task_name in model.prompts:
+            kwargs["prompt_name"] = task_name
+        elif prompt_type and prompt_type in model.prompts:
+            kwargs["prompt_name"] = prompt_type
```

Comment:
It definitely feels a bit hacky but since SentenceTransformers offer a lot of flexibility with the prompt dictionary, I think this is something that we could handle on a potential refactor of `encode_corpus` / `encode_queries` / `encode_conversations` and I'd be happy to help.

Of course, if it's implemented upstream in ST, our life becomes easier, so I'm all for that solution too if Tom deems it appropriate.